enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,320 @@
+perform the upgrade. Such an automation will be part of a future enhancement.
+
+### Workflow Description
+
+1. An engineer working in a partner's factory is asked to prepare a package to upgrade a cluster to
+a specific OpenShift architecture and version.
+
+1. The engineer uses the documentation and tools to prepare the upgrade package containing all the
+artifacts (mostly release images) that are required to perform the upgrade and writes it to some
+kind of media (USB stick, for example) that will be delivered to the technicians.
+
+1. The technicians receive copies of the upgrade package.
+
+1. The technician goes to the cluster location and uses the upgrade package, documentation and tools
+to perform the upgrade. This step is potentially repeated multiple times for multiple clusters.
+
+Note that the upgrade package should not be specific for a particular cluster, only for the
+OpenShift architecture and version. Technicians should be able to use that package for any cluster
+with that architecture and version.
+
+### API Extensions
+
+None.
+
+### Implementation Details/Notes/Constraints
+
+The proposed solution is based on copying all the required images to all the nodes of the cluster
+before starting the upgrade, and ensuring that no component requires access to a registry server
+during the upgrade. For this to work the following changes are required:
+
+1. No OpenShift component used during the upgrade should use the `Always` pull policy, as that
+forces the kubelet and CRI-O to try to contact the registry server even if the image is already
+available.
+
+1. No OpenShift component should garbage collect the upgrade images before or during the upgrade.
+
+1. The machine config operator should tolerate the changes that will be required to the CRI-O
+configuration during the upgrade.
+
+1. The engineer working in the partner's factory needs a supported and documented procedure to
+create the upgrade package.
+
+1. The technician needs a documented and supported procedure to verify that the upgrade requirements
+are met (right version, enough space available, etc) and then to actually perform the upgrade.
+
+#### Don't use the `Always` pull policy during the upgrade
+
+Some OCP core components currently use the `Always` image pull policy during the upgrade. As a
+result, the kubelet and CRI-O will try to contact the registry server even if the image is already
+available in the local storage of the cluster. This blocks the upgrade.
+
+Most OCP core components have been changed in the past to avoid this. Recently the OVN pre-puller
+has also been changed (see this [bug](https://issues.redhat.com/browse/OCPBUGS-13219) for details).
+To prevent bugs like this from happening in the future and make the solution less fragile we should
+have a test that gates the OpenShift release and that verifies that the upgrade can be performed
+without a registry server.
```

Comment:
One way to ensure this is to have an admission hook that rejects/warns any spec with `Always` and run in CI to catch it.

Reply:
Done.